_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+
+```cpp
+void EmployeeTable::addEmployee(const Employee& employee) {
+    employees.push_back(employee);
+}
+
+void EmployeeTable::removeEmployee(const std::string& employeeName) {
+    for (auto it = employees.begin(); it != employees.end(); ++it) {
+        if (it->getName() == employeeName) {
+            employees.erase(it);
+            break;
+        }
+    }
+}
+
+void EmployeeTable::displayEmployeesName() const {
+    cout << "-------------------------------------------------- " << endl;
+    for (const auto& employee : employees) {
+        std::cout << employee.getName() << std::endl;
+    }
+    cout << "-------------------------------------------------- " << endl;
+}
+
+bool EmployeeTable::isEmpty() const {
+    return employees.empty();
+}
+
+int EmployeeTable::getEntryCount() const {
+    return employees.size();
+}
+```
+
+We want to test our table class. In particular, we are interested in testing the following:
+
+1. Table is not empty after adding an employee.
+2. Number of entries is one after adding an employee.
+3. Number of entries in table reduces by one after removing an employee (assuming that there was at least one entry in the table).
+
+Using our knowledge of test fixtures learnt in previous subsection, we can write the tests as shown below. Please see the file [3_emp_table_test.cpp](../code/Chapter3/3_emp_table_test.cpp) for more details.
+
+```cpp
+// Test fixture for EmployeeTable class.
+class EmployeeTableTest : public testing::Test {
+    public:
+        EmployeeTable table;
+};
+
+// Test that the table is not empty after adding an employee.
+TEST_F(EmployeeTableTest, TableIsNotEmptyAfterAddingEmployee) {
+    Employee new_employee("John Doe", 30, 5000, 5, 1000);
+    table.addEmployee(new_employee);
+    EXPECT_FALSE(table.isEmpty());
+}
+
+// Test that number of entries is one after adding an employee.
+TEST_F(EmployeeTableTest, NumberOfEntriesIsOneAfterAddingEmployee) {
+    Employee new_employee("John Doe", 30, 5000, 5, 1000);
+    table.addEmployee(new_employee);
+    EXPECT_EQ(table.getEntryCount(), 1);
+}
+
+// Test that number of entries in table reduces by one after removing an employee.
+TEST_F(EmployeeTableTest, NumberOfEntriesIsOneLessAfterRemovingEmployee) {
+    Employee new_employee("John Doe", 30, 5000, 5, 1000);
+    table.addEmployee(new_employee);
+    table.removeEmployee("John Doe");
+    EXPECT_EQ(table.getEntryCount(), 0);
+}
+```
+
+As we can see from above, that all creating an instance of `employee` first by using the statement like `Employee new_employee` followed by adding an entry to the table `table.addEmployee(new_employee)`. Thus, we can see that our tests need some setup which is common for all and hence `Setup` function comes to rescue for exactly such scenarios.
```

Comment:
I don't understand this sentence

Reply:
Thanks for pointing that out. I missed a few words while writing. I have now corrected it in 94197d78b633256f1a08b783972255fcd130d9a1.